src/engine.rs

```diff
@@ -264,47 +267,64 @@ impl Engine {
+            self.condition.volume,
+            self.condition.fperiod,
+        );
+
         let models = Models::new(
-            labels.labels().to_vec(),
+            labels.labels(),
             &self.voices,
             &self.condition.interporation_weight,
         );
 
+        let estimator = DurationEstimator::new(models.duration(), models.nstate());
         let durations = if self.condition.phoneme_alignment_flag {
-            DurationEstimator.create_with_alignment(&models, labels.times())
+            estimator.create_with_alignment(labels.times())
         } else {
-            DurationEstimator.create(&models, self.condition.speed)
-        };
-
-        let initialize = |stream_index: usize| {
-            MlpgAdjust::new(
-                stream_index,
-                self.condition.gv_weight[stream_index],
-                self.condition.msd_threshold[stream_index],
-            )
-        };
-
-        let spectrum = initialize(0).create(models.stream(0), &models, &durations);
-        let lf0 = {
-            let mut lf0_params = models.stream(1);
-            apply_additional_half_tone(&mut lf0_params, self.condition.additional_half_tone);
-            initialize(1).create(lf0_params, &models, &durations)
+            estimator.create(self.condition.speed)
         };
-        let lpf = initialize(2).create(models.stream(2), &models, &durations);
 
-        let vocoder = Vocoder::new(
-            models.vector_length(0) - 1,
-            self.condition.stage,
-            self.condition.use_log_gain,
-            self.condition.sampling_frequency,
-            self.condition.fperiod,
-        );
-        let generator = SpeechGenerator::new(
-            self.condition.fperiod,
-            self.condition.alpha,
-            self.condition.beta,
-            self.condition.volume,
-        );
-        generator.synthesize(vocoder, spectrum, lf0, Some(lpf))
+        let spectrum =
+            MlpgAdjust::new(self.condition.gv_weight[0], self.condition.msd_threshold[0]).create(
+                models.stream(0),
+                models.vector_length(0),
+                models.windows(0),
+                models.gv(0),
+                &durations,
+            );
+        let lf0 = MlpgAdjust::new(self.condition.gv_weight[1], self.condition.msd_threshold[1])
+            .create(
+                mutated(models.stream(1), |params| {
+                    apply_additional_half_tone(params, self.condition.additional_half_tone);
+                }),
+                models.vector_length(1),
+                models.windows(1),
+                models.gv(1),
+                &durations,
+            );
+        let lpf = MlpgAdjust::new(self.condition.gv_weight[2], self.condition.msd_threshold[2])
+            .create(
+                models.stream(2),
+                models.vector_length(2),
+                models.windows(2),
+                models.gv(2),
+                &durations,
+            );
```

Comment:
LPFがない場合はpanicするということですか？LPFがないのも正常系なので，panicするには実装が難しいとか，理由をつける必要があると思います．

Reply:
panicしないならこんな感じですかね．
```suggestion
        let lpf = if self.voices.global_metadata().num_streams >= 3 {
            MlpgAdjust::new(self.condition.gv_weight[2], self.condition.msd_threshold[2]).create(
                models.stream(2),
                models.vector_length(2),
                models.windows(2),
                models.gv(2),
                &durations,
            )
        } else {
            vec![vec![0.0; 0]; lf0.len()]
        };
```